src/AsmResolver.PE.Win32Resources/Icon/IconResource.cs

```diff
@@ -0,0 +1,142 @@
+        /// <param name="id">The id of the icon group.</param>
+        public IconGroupDirectory this[uint id]
+        {
+            get => _entries[id];
+            set
+            {
+                if (value is null)
+                    _entries.Remove(id);
+                else
+                    _entries[id] = value;
+            }
+        }
+
+        /// <summary>
+        /// Adds or overrides the existing entry with the same id to the icon group resource. 
+        /// </summary>
+        /// <param name="id">The id to use for the entry.</param>
+        /// <param name="entry">The entry to add.</param>
+        public void AddEntry(uint id, IconGroupDirectory entry) => _entries[id] = entry;
+
+        /// <summary>
+        /// Gets a collection of entries stored in the icon group directory. 
+        /// </summary>
+        /// <returns>The collection of icon group entries.</returns>
+        public IEnumerable<IconGroupDirectory> GetIconGroups() => _entries.Values;
+
+        /// <inheritdoc />
+        public void WriteToDirectory(IResourceDirectory rootDirectory)
+        {
+            // Find and remove old group icon directory.
+            int groupIconIndex = IndexOfResourceDirectoryType(rootDirectory, ResourceType.GroupIcon);
+
+            if (groupIconIndex == -1)
+                groupIconIndex = rootDirectory.Entries.Count;
+            else
+                rootDirectory.Entries.RemoveAt(groupIconIndex);
+
+            // Find and remove old icon directory.
+            int iconIndex = IndexOfResourceDirectoryType(rootDirectory, ResourceType.Icon);
+
+            if (iconIndex == -1)
+                iconIndex = rootDirectory.Entries.Count;
+            else
+                rootDirectory.Entries.RemoveAt(iconIndex);
+
+            if (groupIconIndex == iconIndex)
+                iconIndex += 1;
+
+            // Construct new directory.
+            var newGroupIconDirectory = new ResourceDirectory(ResourceType.GroupIcon);
+            foreach (var entry in _entries)
+            {
+                newGroupIconDirectory.Entries.Add(new ResourceDirectory(entry.Key)
+                    {Entries = {new ResourceData(0u, entry.Value)}});
+            }
+
+            // Construct new directory.
+            var newIconDirectory = new ResourceDirectory(ResourceType.Icon);
+            foreach (var entry in _entries)
+            {
+                foreach (var icon in entry.Value.GetIconEntries())
+                {
+                    newIconDirectory.Entries.Add(new ResourceDirectory(icon.Item1.Id)
+                        {Entries = {new ResourceData(0u, icon.Item2)}});
+                }
+            }
+
+            // Insert.
+            rootDirectory.Entries.Insert(groupIconIndex, newGroupIconDirectory);
+            rootDirectory.Entries.Insert(iconIndex, newIconDirectory);
+        }
+
+        private static int IndexOfResourceDirectoryType(IResourceDirectory rootDirectory, ResourceType resourceType)
```

Comment:
This is a duplicate of the one used in `VersionInfoResource`. We probably should reuse it instead and maybe move it to some kind of `internal static Utilities` class.